server/src/main/java/org/elasticsearch/node/InternalSettingsPreparer.java

```diff
@@ -113,6 +108,59 @@ static void initializeSettings(final Settings.Builder output, final Settings inp
         output.replacePropertyPlaceholders();
     }
 
+    static void loadConfigWithSubstitutions(Settings.Builder output, Path configFile, Function<String, String> substitutions)
+        throws IOException {
+        long existingSize = Files.size(configFile);
+        StringBuilder builder = new StringBuilder((int) existingSize);
+        try (BufferedReader reader = Files.newBufferedReader(configFile, StandardCharsets.UTF_8)) {
+            String line;
+            while ((line = reader.readLine()) != null) {
+                int dollarNdx;
+                int nextNdx = 0;
+                while ((dollarNdx = line.indexOf("${", nextNdx)) != -1) {
+                    int closeNdx = line.indexOf('}', dollarNdx + 2);
+                    if (closeNdx == -1) {
+                        // No close substitution was found. Break to leniently copy the rest of the line as is.
+                        break;
+                    }
+                    // copy up to the dollar
+                    if (dollarNdx > nextNdx) {
+                        builder.append(line, nextNdx, dollarNdx);
+                    }
+                    nextNdx = closeNdx + 1;
+
+                    String substKey = line.substring(dollarNdx + 2, closeNdx);
+                    String substValue = substitutions.apply(substKey);
+                    if (substValue != null) {
+                        builder.append(substValue);
+                    } else {
+                        // the substitution name doesn't exist, defer to setting based substitution after yaml parsing
+                        builder.append(line, dollarNdx, nextNdx);
+                    }
+                }
+                if (nextNdx < line.length()) {
+                    builder.append(line, nextNdx, line.length());
+                }
+                builder.append(System.lineSeparator());
+            }
+        }
+        var is = new ByteArrayInputStream(builder.toString().getBytes(StandardCharsets.UTF_8));
+        output.loadFromStream(configFile.getFileName().toString(), is, false);
+    }
+
+    static void loadOverrides(Settings.Builder output, Map<String, String> overrides) throws IOException {
+        StringBuilder builder = new StringBuilder();
+        for (var entry : overrides.entrySet()) {
+            builder.append(entry.getKey());
+            builder.append(": ");
+            builder.append(entry.getValue());
+            builder.append(System.lineSeparator());
+        }
+        var is = new ByteArrayInputStream(builder.toString().getBytes(StandardCharsets.UTF_8));
+        // fake the resource name so it loads yaml
+        output.loadFromStream("overrides.yml", is, false);
```

Comment:
I'm confused why this is necessary. Doesn't the later call to  `initializeSettings` take care of replacing any remaining values? How many "passes" are necessary?

Reply:
Oops, I meant to remove the putProperties in initializeSettings (I've done so now). The idea is that instead of putting the values directly (which may be json), we load them here as if they set in a yaml file.